terraform/node_resource_abstract_instance.go

```diff
@@ -812,6 +813,35 @@ func (n *NodeAbstractResourceInstance) plan(
 		}
 	}
 
+	// The user might also ask us to force replacing a particular resource
+	// instance, regardless of whether the provider thinks it needs replacing.
+	// For example, users typically do this if they learn a particular object
+	// has become degraded in an immutable infrastructure scenario and so
+	// replacing it with a new object is a viable repair path.
+	matchedForceReplace := false
+	for _, candidateAddr := range forceReplace {
+		if candidateAddr.Equal(n.Addr) {
+			matchedForceReplace = true
+			break
+		}
+		// For "force replace" purposes we require an exact resource instance
+		// address to match, but just in case a user forgets to include the
+		// instance key for a multi-instance resource we'll give them a
+		// warning hint.
+		if n.Addr.Resource.Key != addrs.NoKey && candidateAddr.Resource.Key == addrs.NoKey {
```

Comment:
I liked doing this down here because it allows the error message to include a precise example of what to run to get the intended result, rather than making the user figure out themselves what the instance keys for this resource might be. But I agree it does seem quite bulky.

In principle we could check it earlier in the pre-expansion graph node and generate one message that lists out N separate options to add, but my assumption here is that the primary use-case for this option is that _one_ object has become degraded and needs replacing, and so it felt weird to encourage just unilaterally replacing the whole set rather than picking out just the one you intended to select.

:thinking: I'm not sure either. What do you think about running with this implementation for now and then considering adjustments once we see how this option gets used? Hopefully some real-world feedback will help evaluate what cases are most common here and thus what the best UI might be, and specific error message text is never a compatibility constraint so we can freely change it over time as we learn more.

Reply:
Hrm, I see what you're saying. My _feeling_ is that anything more than one message is too many, and anyone with, say, 3 or more resource instances will end up with too much output to get a useful signal. I would rather see an earlier message that said  "here are all of the instances of this resource address, pick one", but I'm also amenable to your suggestion that we wait and see what feedback we get.

Reply:
I'd have to see what the actual output is, but I have 2 suspicions about what it will look like... the user misses one instance key and gets a wall of diagnostic showing lots of instances they don't care about, or the diagnostics are collapsed and they get a message about one random instance they don't care about. 

I think it may be better to do this earlier and craft a message about the address with a list of possible keys. We can also run with this to get the implementation in place and change it later if we want.

Reply:
Emitting a single message earlier would mean doing this check again in another place, which is fine but I'm inclined to save it for a later enhancement because while I'm still working through this it's likely I'll be fiddling with these details some more before we reach the end of this PR sequence and I'd rather only churn one place. I'll return to this once I get stuck into the UI work for both of these and see about adding a new check in a different graph node.